File: CODIGO7/exercise.scala
object LuckyTickets {
  def luckyTickets(n: Int, b: Int): Int = {
    var luckyTickets = 0
    for (i <- 0 until (1 << n) - 1) {
      var firstHalf = 0
      var secondHalf = 0
      var j = 0
      while (i > 0) {
        firstHalf += i % b if j < n / 2 else secondHalf += i % b
        i >>= 1
        j += 1
      }
      if (firstHalf == secondHalf) {
        luckyTickets += 1
      }
    }
    luckyTickets
  }

  def main(args: Array[String]) {
    val n = scala.io.StdIn.readInt()
    val b = scala.io.StdIn.readInt()
    println(luckyTickets(n, b))
  }
}

File: CODIGO7/exercise.rb
def lucky_tickets(n, b)
    lucky_tickets = 0
    max_value = (1 << n) - 1
    for i in 0..max_value
      first_half = 0
      second_half = 0
      j = 0
      while i > 0
        digit = i % b
        first_half += digit if j < n / 2
        second_half += digit if j >= n / 2
        i >>= 1
        j += 1
      end
      lucky_tickets += 1 if first_half == second_half
    end
    return lucky_tickets
  end
  
  def main()
    n, b = gets.split(' ').map(&:to_i)
    puts lucky_tickets(n, b)
  end
  
  main()
  